### PR TITLE
fix-cache-returns-nil

### PR DESCRIPTION
**Pull request by masarakki**
In my instance, sometimes `user.settings.interactions` returns `nil` .
I understand that this code never returns `nil`, but something happen...

### CODE REVIEW

**Comment by nightpool** [comment]
Looks like this changes a bunch of existing behavior—see the test failures. Can you outline exactly what the issue is and why it's a problem?

If you say "that this code never returns nil", then you should probably investigate *why* it's going on not just masking the symptoms.

**Comment by Gargron** [comment]
Does `user.settings.interactions` for the same user sometimes return nil, or is it more of an issue of only some users missing defaults?

**Comment by masarakki** [comment]
this happens both case user has some settings / no settings ,

I found what happened on rails console,
```
irb(main):048:0> Rails.cache.exist?(Setting.cache_key('interactions', user))
=> true
irb(main):049:0> Rails.cache.fetch(Setting.cache_key('interactions', user))
=> nil
# ..few minutes after
irb(main):050:0> Rails.cache.fetch(Setting.cache_key('interactions', user))
=> {"must_be_follower"=>false, "must_be_following"=>false}
```

**Comment by Gargron** [comment]
So the problem is the old result is cached and the cache is not discarded when setting is updated? Or am I still misunderstanding what happens and why?

**Comment by masarakki** [comment]
Essentially, this problem is `db_val` of 'interactions' or `defuallt_values['interactions']` never returns `nil`, but (it seems that) redis stored `nil`

File: app/models/setting.rb [hunk 1]
```diff
@@ -35,7 +36,7 @@ def [](key)
           default_settings[key]
         end
       end
-
+      val = default_settings[key] unless Rails.cache.exist?(key)
```
Comment (Gargron) [inline app/models/setting.rb]: Should this not be `if val.nil?`

**Comment by masarakki** [comment]
```
rails c                                [19:13:37]
Loading development environment (Rails 5.0.2)
[1] pry(main)> Settings::ScopedSettings.instance_variable_get(:@object)
=> nil
[2] pry(main)> User.first.settings
  User Load (0.6ms)  SELECT  "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> Settings::ScopedSettings(id: integer, var: string, value: text, thing_type: string, thing_id: integer, created_at: datetime, updated_at: datetime)
[3] pry(main)> Settings::ScopedSettings.instance_variable_get(:@object)
=> #<User id: 1, email:`...>
```

I think it is the factor of its problem.
I fix to [using instance of `ScopeSettings`](https://github.com/dwango/mastodon/pull/68), and I'm testing in my instance.

**Comment by alpaca-tc** [comment]
I noticed that current `Settings::ScopedSettings` is not thread-safe. I might be wrong, but is this the cause? 🤔 
I am going to create new PR to fix it.

**Comment by Gargron** [comment]
@alpaca-tc Perhaps we should add monitor/synchronize to it? I don't know how to reproduce this issue for testing, so I am just guessing...

**Comment by alpaca-tc** [comment]
@Gargron I am just guessing, too. I will try to reproduce it. 😕

**Comment by alpaca-tc** [comment]
@Gargron just finished, we should fix this problem as soon as possible.
I'll try to create pull-request tomorrow 😄 

```
# evaluate script on rails console or rails runner
ActiveRecord::Base.logger = Logger.new('/dev/null')
raise 'please enable redis store' unless Rails.cache.is_a?(ActiveSupport::Cache::RedisStore)

def update_settings(user)
  new_value = user.settings['test'].to_i + 1
  user.settings['test'] = new_value # UPDATE "settings" SET "value" = $1, "updated_at" = $2 WHERE "settings"."id" = $3
end

begin
  threads = []

  # This thread doesn't update settings.
  # If other thread updates record or cache, this thread raise error.
  threads << Thread.new do
    user = User.find(1)
    previous_value = user.settings['test']

    50.times { user.settings; sleep(0.1) } # Overwrite Settings::ScopedSettings's @object to check thread-safe

    current_value = user.settings['test']

    unless previous_value == current_value
      raise "Ooops!! settings is updated by other thread. prevous:#{previous_value}, current:#{current_value}" 
    end
  end

  # Unfortunately, this thread updates settings of first user
  threads << Thread.new { 50.times { update_settings(User.find(2)) } }

  threads.each(&:join)
rescue => error
  puts error
end
```

**Comment by alpaca-tc** [comment]
Ah, it seems that this PR fixed the problem, does not? @masarakki 

I hope to merge after our reviewing. how about it? @Gargron

**Review by Gargron — APPROVED** [review]

**Comment by Gargron** [comment]
Yes I didn't notice the refactor was already in the PR. Still, should I wait for you to confirm this fixes the issue, @alpaca-tc ?

**Comment by masarakki** [comment]
After this patch,  `Setting` is (maybe) used only for global settings.
So I think it's not so dangerous.

I suggest to avoid using `rails-settings`.

**Comment by Gargron** [comment]
I am adding a Redis lock gem in #4258. One of the examples in the gem documentation is using the lock for fetching cache: <https://github.com/marioizquierdo/mario-redis-lock/blob/master/EXAMPLE_DOG_PILE_EFFECT.md> Perhaps this would be useful for this code? I think it's still a bit buggy, I remember sometimes seeing in my sidekiq logs that some local user's settings which should never be nil return nil and the job fails

**Pull request merged**